File: materials/nuclear_data_manager.py
import json
import os
from typing import Dict, List, Optional, Tuple
from .models import ElementData, IsotopeData, SubstitutionRule

class NuclearDataManager:
    """
    Singleton for managing nuclear data (atomic masses, abundances, substitutions).
    """
    _instance = None
    
    def __new__(cls, *args, **kwargs):
        if not cls._instance:
            cls._instance = super(NuclearDataManager, cls).__new__(cls)
            cls._instance._initialized = False
        return cls._instance

    def __init__(self, data_dir: str = None):
        if self._initialized:
            return
            
        if data_dir is None:
            # 1. Try environment variable
            data_dir = os.environ.get("NUCLEAR_DATA_PATH")
            
        if data_dir is None:
            # 2. Try package installation path
            base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
            # If installed as package, nuclear_data is likely at base_dir/nuclear_data
            data_dir = os.path.join(base_dir, "nuclear_data")
            
        if not os.path.exists(data_dir):
            raise FileNotFoundError(f"Nuclear data directory not found at: {data_dir}")
            
        self.data_dir = data_dir
        self.elements: Dict[str, ElementData] = {}
        self.periodic_table: Dict[str, int] = {}
        self.z_to_symbol: Dict[int, str] = {}
        self.substitutions: Dict[int, SubstitutionRule] = {}
        self.atomic_masses: Dict[Tuple[int, int], float] = {}
        
        self.load_all()
        self._initialized = True

    def load_all(self):
        """Load all JSON data files."""
        self._load_periodic_table()
        self._load_natural_abundances()
        self._load_atomic_masses()
        self._load_substitutions()

    def _load_periodic_table(self):
        path = os.path.join(self.data_dir, "periodic_table.json")
        with open(path, 'r') as f:
            self.periodic_table = json.load(f)
            self.z_to_symbol = {v: k for k, v in self.periodic_table.items()}

    def _load_natural_abundances(self):
        path = os.path.join(self.data_dir, "natural_abundances.json")
        with open(path, 'r') as f:
            data = json.load(f)
            for sym, info in data.items():
                isotopes = [IsotopeData(**iso) for iso in info['isotopes']]
                self.elements[sym] = ElementData(
                    Z=info['Z'],
                    symbol=sym,
                    isotopes=isotopes,
                    collprob=info['collprob']
                )

    def _load_atomic_masses(self):
        path = os.path.join(self.data_dir, "atomic_masses.json")
        with open(path, 'r') as f:
            data = json.load(f)
            for sym, info in data.items():
                Z = info['Z']
                for iso in info['isotopes']:
                    self.atomic_masses[(Z, iso['A'])] = iso['mass']

    def _load_substitutions(self):
        path = os.path.join(self.data_dir, "xs_substitutions.json")
        with open(path, 'r') as f:
            data = json.load(f)
            for sub in data.get("substitutions", []):
                from_zaid = self.parse_zaid(sub["from"])
                to_zaid = self.parse_zaid(sub["to"]) if sub.get("to") else None
                rule = SubstitutionRule(from_zaid=from_zaid, to_zaid=to_zaid, reason=sub.get("reason", ""))
                self.substitutions[from_zaid] = rule

    def parse_zaid(self, zaid_str: str) -> int:
        """Parse 'O-18' or '8018' to integer ZAID."""
        if '-' in zaid_str:
            sym, a_str = zaid_str.split('-')
            z = self.periodic_table.get(sym)
            if z is None:
                raise ValueError(f"Unknown element symbol: {sym}")
            return z * 1000 + int(a_str)
        return int(zaid_str)

    def get_element(self, symbol: str) -> ElementData:
        return self.elements.get(symbol)

    def get_mass(self, z: int, a: int) -> float:
        return self.atomic_masses.get((z, a))

    def get_symbol(self, z: int) -> str:
        return self.z_to_symbol.get(z)

    def get_z(self, symbol: str) -> int:
        return self.periodic_table.get(symbol)

File: cppapi/nuclear_data_manager.cpp
export module materiais.nuclear_data_manager;

#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <string>
#include <map>
#include <array>
#include <vector>

import materiais.models;
import materiais.json;

export namespace materiais {

NuclearDataManager::NuclearDataManager(std::string data_dir)
    : data_dir_(std::move(data_dir)) {
    if (data_dir_.empty()) {
        data_dir_ = std::string(__FILE__);
        auto pos = data_dir_.find_last_of("/");
        if (pos != std::string::npos) {
            data_dir_.erase(pos + 1);
        }
        data_dir_ += "../nuclear_data";
    }
    load_all();
}

const ElementData* NuclearDataManager::get_element(const std::string& symbol) const noexcept {
    auto it = elements_.find(symbol);
    return it == elements_.end() ? nullptr : &it->second;
}

double NuclearDataManager::get_mass(int z, int a) const noexcept {
    int key = z * 1000 + a;
    auto it = atomic_masses_.find(key);
    return it == atomic_masses_.end() ? 0.0 : it->second;
}

std::string NuclearDataManager::get_symbol(int z) const noexcept {
    if (z > 0 && z < static_cast<int>(periodic_table_.size())) {
        return periodic_table_[z];
    }
    return {};
}

int NuclearDataManager::get_z(const std::string& symbol) const noexcept {
    for (int index = 1; index < static_cast<int>(periodic_table_.size()); ++index) {
        if (periodic_table_[index] == symbol) {
            return index;
        }
    }
    return 0;
}

const SubstitutionRule* NuclearDataManager::get_substitution(int from_zaid) const noexcept {
    auto it = substitutions_.find(from_zaid);
    return it == substitutions_.end() ? nullptr : &it->second;
}

void NuclearDataManager::load_all() {
    load_periodic_table();
    load_natural_abundances();
    load_atomic_masses();
    load_substitutions();
}

void NuclearDataManager::load_periodic_table() {
    auto root = parse_json_file(data_dir_ + "/periodic_table.json");
    if (!root.is_object()) {
        throw std::runtime_error("Invalid periodic_table.json");
    }
    for (const auto& pair : root.as_object()) {
        const std::string& symbol = pair.first;
        const JsonValue& entry = pair.second;
        int z = 0;
        if (entry.is_number()) {
            z = static_cast<int>(entry.as_number());
        } else if (entry.is_object()) {
            z = static_cast<int>(entry.as_object().at("Z").as_number());
        }
        if (z > 0 && z < static_cast<int>(periodic_table_.size())) {
            periodic_table_[z] = symbol;
        }
    }
}

void NuclearDataManager::load_natural_abundances() {
    auto root = parse_json_file(data_dir_ + "/natural_abundances.json");
    if (!root.is_object()) {
        throw std::runtime_error("Invalid natural_abundances.json");
    }
    for (const auto& pair : root.as_object()) {
        const std::string& symbol = pair.first;
        const auto& entry = pair.second.as_object();
        ElementData element;
        element.Z = static_cast<int>(entry.at("Z").as_number());
        element.symbol = symbol;
        for (const auto& item : entry.at("isotopes").as_array()) {
            const auto& iso = item.as_object();
            IsotopeData data;
            data.A = static_cast<int>(iso.at("A").as_number());
            data.mass = iso.at("mass").as_number();
            data.abundance = iso.at("abundance").as_number();
            element.isotopes.push_back(data);
        }
        for (const auto& value : entry.at("collprob").as_array()) {
            element.collprob.push_back(value.as_number());
        }
        elements_.emplace(symbol, std::move(element));
    }
}

void NuclearDataManager::load_atomic_masses() {
    auto root = parse_json_file(data_dir_ + "/atomic_masses.json");
    if (!root.is_object()) {
        throw std::runtime_error("Invalid atomic_masses.json");
    }
    for (const auto& pair : root.as_object()) {
        const auto& entry = pair.second.as_object();
        int z = static_cast<int>(entry.at("Z").as_number());
        for (const auto& item : entry.at("isotopes").as_array()) {
            const auto& iso = item.as_object();
            int a = static_cast<int>(iso.at("A").as_number());
            int key = z * 1000 + a;
            atomic_masses_[key] = iso.at("mass").as_number();
        }
    }
}

void NuclearDataManager::load_substitutions() {
    auto root = parse_json_file(data_dir_ + "/xs_substitutions.json");
    if (!root.is_object()) {
        return;
    }
    const auto& array = root.as_object().at("substitutions").as_array();
    for (const auto& item : array) {
        const auto& entry = item.as_object();
        int from_zaid = parse_zaid(entry.at("from").as_string());
        int to_zaid = -1;
        if (!entry.at("to").is_null()) {
            to_zaid = parse_zaid(entry.at("to").as_string());
        }
        SubstitutionRule rule;
        rule.from_zaid = from_zaid;
        rule.to_zaid = to_zaid;
        if (!entry.at("reason").is_null()) {
            rule.reason = entry.at("reason").as_string();
        }
        substitutions_.emplace(from_zaid, std::move(rule));
    }
}

int NuclearDataManager::parse_zaid(const std::string& zaid_str) const {
    size_t sep = zaid_str.find('-');
    if (sep != std::string::npos) {
        std::string symbol = zaid_str.substr(0, sep);
        int A = std::stoi(zaid_str.substr(sep + 1));
        int Z = get_z(symbol);
        return Z * 1000 + A;
    }
    return std::stoi(zaid_str);
}

} // namespace materiais
